pvlib/ivtools.py

```diff
@@ -262,6 +262,173 @@ def fit_sde_sandia(voltage, current, v_oc=None, i_sc=None, v_mp_i_mp=None,
+        from scipy.optimize import root
+        from scipy import constants
+    except ImportError:
+        raise ImportError("The fit_sdm_desoto function requires scipy.")
+
+    # Constants
+    k = constants.value('Boltzmann constant in eV/K')
+    Tref = temp_ref + 273.15  # [K]
+
+    def pv_fct(params, specs):
+        """Evaluates the systems of equations used to solve for the single
+        diode equation parameters.
+        To avoid the confusion in names with variables of container
+        function the '_' of the variables were removed.
+        """
+        # six input known variables
+        Isc, Voc, Imp, Vmp, betaoc, alphasc = specs
+
+        # five parameters vector to find
+        IL, Io, a, Rsh, Rs = params
+
+        # five equation vector
+        y = [0, 0, 0, 0, 0]
+
+        # 1st equation - short-circuit - eq(3) in [1]
+        y[0] = Isc - IL + Io*np.expm1(Isc*Rs/a) + Isc*Rs/Rsh
+        # 2nd equation - open-circuit Tref - eq(4) in [1]
+        y[1] = -IL + Io*np.expm1(Voc/a) + Voc/Rsh
+        # 3rd equation - Imp & Vmp - eq(5) in [1]
+        y[2] = Imp - IL + Io*np.expm1((Vmp+Imp*Rs)/a) + \
+            (Vmp+Imp*Rs)/Rsh
+        # 4th equation - Pmp derivated=0 - eq23.2.6 in [2]
+        # caution: eq(6) in [1] has a sign error
+        y[3] = Imp - Vmp * ((Io/a)*np.exp((Vmp+Imp*Rs)/a) + 1.0/Rsh) / \
+            (1.0 + (Io*Rs/a)*np.exp((Vmp+Imp*Rs)/a) + Rs/Rsh)
+        # 5th equation - open-circuit T2 - eq (4) at temperature T2 in [1]
+        T2 = Tref + 2
+        Voc2 = (T2 - Tref)*betaoc + Voc  # eq (7) in [1]
+        a2 = a*T2/Tref  # eq (8) in [1]
+        IL2 = IL + alphasc*(T2-Tref)  # eq (11) in [1]
+        Eg2 = EgRef*(1 + dEgdT*(T2-Tref))  # eq (10) in [1]
+        Io2 = Io * (T2/Tref)**3 * np.exp(1/k * (EgRef/Tref-Eg2/T2))  # eq (9)
+        y[4] = -IL2 + Io2*np.expm1(Voc2/a2) + Voc2/Rsh  # eq (4) at T2
+
+        return y
+
+    # initial guesses of variables for computing convergence:
+    # Values are taken from [2], p753
+    Rsh_i = 100.0
+    a_i = 1.5*k*Tref*cells_in_series
+    IL_i = i_sc
+    Io_i = i_sc * np.exp(-v_oc/a_i)
+    Rs_i = (a_i*np.log1p((IL_i-i_mp)/Io_i) - v_mp)/i_mp
+    # params_i : initial values vector
+    params_i = np.array([IL_i, Io_i, a_i, Rsh_i, Rs_i])
+
+    # specs of module
+    specs = np.array([i_sc, v_oc, i_mp, v_mp, beta_voc, alpha_sc])
+
+    # computing
+    result = root(pv_fct, x0=params_i, args=specs)
+
+    if result.success:
+        sdm_params = result.x
+    else:
+        raise RuntimeError('Parameter estimation failed:\n' + result.message)
+
+    # results
+    return {'I_L_ref': sdm_params[0],
+            'I_o_ref': sdm_params[1],
+            'a_ref': sdm_params[2],
+            'R_sh_ref': sdm_params[3],
+            'R_s': sdm_params[4],
+            'alpha_sc': alpha_sc,
+            'EgRef': EgRef,
+            'dEgdT': dEgdT,
+            'irrad_ref': irrad_ref,
+            'temp_ref': temp_ref,
+            'optimize_result': result}
```

Comment:
Unfortunately, including the `optimize_result` right in this dictionary makes chaining computations more difficult:
```python
result = ivtools.fit_sdm_desoto(cells_in_series=60, v_mp=31.0, i_mp=8.71, v_oc=38.3, i_sc=9.43, alpha_sc=0.005658, beta_voc=-0.13788)
pvlib.pvsystem.calcparams_desoto(1000, 25, **result)
```
```console
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-17-57980407c8d1> in <module>
----> 1 pvlib.pvsystem.calcparams_desoto(1000, 25, **result)

TypeError: calcparams_desoto() got an unexpected keyword argument 'optimize_result'
```
There are at least a few options here:
1. Ignore my suggestion to include the `optimize_result`.
2. Take my original suggestion of making `optimize_result` the second element in a returned tuple.
  ```python
      return {'I_L_ref': sdm_params[0], 'I_o_ref': sdm_params[1], 'a_ref': sdm_params[2], 'R_sh_ref': sdm_params[3],
                  'R_s': sdm_params[4], 'alpha_sc': alpha_sc, 'EgRef': EgRef, 'dEgdT': dEgdT, 'irrad_ref': irrad_ref,
                  'temp_ref': temp_ref}, result
  ```
3. Make the returned dictionary have a nested key for the parameters, such as:
  ```python
      return {
                  'params': {
                                  'I_L_ref': sdm_params[0],
                                  'I_o_ref': sdm_params[1],
                                  'a_ref': sdm_params[2],
                                  'R_sh_ref': sdm_params[3],
                                  'R_s': sdm_params[4],
                                  'alpha_sc': alpha_sc,
                                  'EgRef': EgRef,
                                  'dEgdT': dEgdT,
                                  'irrad_ref': irrad_ref,
                                  'temp_ref': temp_ref
                                 },
                  'optimize_result': result
  }
  ```

Reply:
I apologize, I overlooked this change.  _If_ we are returning the `result` object, return as 2nd element in a tuple.  `optimize_result` would be a better variable name.

Reply:
Sorry, I made this change too quickly.
@markcampanelli I think your view on the package is more complete than mine, so I will include the optimize_result as you suggested. (after all, the goal of this pull request is also to learn from experienced users/contributors of pvlib-python)
@cwhanse Why do you prefer to return it in a tuple? Intuitively I would have use the dict (3rd option of markcampanelli) for improving the readability in the later use of the output.

Reply:
so that it can be omitted if not of interest, e.g., `params, _ = fit_sdm_desoto(`

Reply:
Like option 2. I don't like option 3.

Reply:
Thanks for chiming in @cwhanse and @wholmgren. I'm good with option 2. Interestingly, @tylunel, I use option 3 in my own APIs (I always return a single dictionary, nested if necessary), but for some reason I guessed that option 2 would be more preferable than option 3 in pvlib ;).